src/dbnode/encoding/proto/encoder.go

```diff
@@ -0,0 +1,820 @@
+					return fmt.Errorf(
+						"error: %v, checking if %v is default value for field %s",
+						err, curVal, field.String())
+				}
+				if isDefaultValue {
+					fieldsChangedToDefault = append(fieldsChangedToDefault, fieldNum)
+				}
+
+				changedFields = append(changedFields, fieldNum)
+				if err := enc.lastEncoded.TrySetFieldByNumber(fieldNumInt, curVal); err != nil {
+					return fmt.Errorf(
+						"error: %v setting field %d with value %v on lastEncoded",
+						err, fieldNumInt, curVal)
+				}
+			}
+		}
+	}
+
+	if len(changedFields) == 0 && enc.lastEncoded != nil {
+		// Only want to skip encoding if nothing has changed AND we've already
+		// encoded the first message.
+		enc.stream.WriteBit(opCodeNoChange)
+		return nil
+	}
+
+	// TODO(rartoul): Need to add a MarshalInto to the ProtoReflect library to save
+	// allocations: https://github.com/m3db/m3/issues/1471
+	marshaled, err := m.Marshal()
+	if err != nil {
+		return fmt.Errorf("%s error trying to marshal protobuf: %v", encErrPrefix, err)
+	}
+
+	// Control bit indicating that proto values have changed.
+	enc.stream.WriteBit(opCodeChange)
+	if len(fieldsChangedToDefault) > 0 {
+		// Control bit indicating that some fields have been set to default values
+		// and that a bitset will follow specifying which fields have changed.
+		enc.stream.WriteBit(opCodeFieldsSetToDefaultProtoMarshal)
+		enc.encodeBitset(fieldsChangedToDefault)
+	} else {
+		// Control bit indicating that none of the changed fields have been set to
+		// their default values so we can do a clean merge on read.
+		enc.stream.WriteBit(opCodeNoFieldsSetToDefaultProtoMarshal)
+	}
+	enc.encodeVarInt(uint64(len(marshaled)))
+	enc.stream.WriteBytes(marshaled)
+
+	if enc.lastEncoded == nil {
+		// Set lastEncoded to m so that subsequent encodings only need to encode fields
+		// that have changed.
+		enc.lastEncoded = dynamic.NewMessage(enc.schema)
+		enc.lastEncoded.Merge(m)
+	} else {
+		// lastEncoded has already been mutated to reflect the current state.
+	}
+
+	return nil
+}
+
+func (enc *Encoder) encodeFirstTSZValue(i int, v float64) {
+	fb := math.Float64bits(v)
+	enc.stream.WriteBits(fb, 64)
+	enc.customFields[i].prevFloatBits = fb
+	enc.customFields[i].prevXOR = fb
+}
+
+func (enc *Encoder) encodeNextTSZValue(i int, next float64) {
+	curFloatBits := math.Float64bits(next)
+	curXOR := enc.customFields[i].prevFloatBits ^ curFloatBits
+	m3tsz.WriteXOR(enc.stream, enc.customFields[i].prevXOR, curXOR)
+	enc.customFields[i].prevFloatBits = curFloatBits
+	enc.customFields[i].prevXOR = curXOR
+}
+
+func (enc *Encoder) encodeFirstSignedIntValue(i int, v int64) {
+	neg := false
+	enc.customFields[i].prevFloatBits = uint64(v)
+	if v < 0 {
+		neg = true
+		v = -1 * v
+	}
+
+	vBits := uint64(v)
+	numSig := encoding.NumSig(vBits)
+
+	m3tsz.WriteIntSig(enc.stream, &enc.customFields[i].intSigBitsTracker, numSig)
+	enc.encodeIntValDiff(vBits, neg, numSig)
+}
+
+func (enc *Encoder) encodeFirstUnsignedIntValue(i int, v uint64) {
```

Comment:
Same comment for firstsignedintvalue and firstunsignedintvalue and the consequent "next" versions too.